GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -53,70 +52,101 @@ public void Dispose()
-            IProductUpgrader productUpgrader;
-            bool deleteExistingDownloads = true;
-
-            if (ProductUpgraderFactory.TryCreateUpgrader(out productUpgrader, this.tracer, out errorMessage))
-            {
-                if (prerunChecker.TryRunPreUpgradeChecks(out string _) && this.TryDownloadUpgrade(productUpgrader, out errorMessage))
-                {
-                    deleteExistingDownloads = false;
-                }
-            }
-
-            if (errorMessage != null)
+            EventMetadata metadata = new EventMetadata();
+            if (e != null)
             {
-                this.tracer.RelatedError(errorMessage);
+                metadata.Add("Exception", e.ToString());
             }
 
-            if (deleteExistingDownloads)
-            {
-                ProductUpgraderInfo.DeleteAllInstallerDownloads();
-            }
+            return metadata;
         }
 
-        private bool TryDownloadUpgrade(IProductUpgrader productUpgrader, out string errorMessage)
+        private void TimerCallback(object unusedState)
         {
+            string errorMessage = null;
+
             using (ITracer activity = this.tracer.StartActivity("Checking for product upgrades.", EventLevel.Informational))
             {
-                Version newerVersion = null;
-                string detailedError = null;
-                if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                try
                 {
-                    return false;
-                }
+                    // The upgrade check always goes against GitHub
+                    GitHubUpgrader productUpgrader = GitHubUpgrader.Create(
+                        this.tracer,
+                        false,
+                        false,
+                        out errorMessage);
 
-                if (!productUpgrader.TryQueryNewestVersion(out newerVersion, out detailedError))
-                {
-                    errorMessage = "Could not fetch new version info. " + detailedError;
-                    return false;
-                }
+                    if (productUpgrader == null)
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
 
-                if (newerVersion == null)
-                {
-                    // Already up-to-date
-                    // Make sure there a no asset installers remaining in the Downloads directory. This can happen if user
-                    // upgraded by manually downloading and running asset installers.
-                    ProductUpgraderInfo.DeleteAllInstallerDownloads();
-                    errorMessage = null;
-                    return true;
-                }
+                        return;
+                    }
 
-                if (productUpgrader.TryDownloadNewestVersion(out detailedError))

```

Comment:
So when do we actually download the packages now? Also what would happen if the version number in the 'latest' file is different from the actual latest version on the server/feed? Do we download the version number as specified in the file or just 'latest'?

(Just thinking what would happen in the case we pulled an update file but a user already did the check and has a 'latest' file pointing to that now revoked version.)

Reply:
> So when do we actually download the packages now?

We download the package when running the Upgrade command.

> Also what would happen if the version number in the 'latest' file is different from the actual latest version on the server/feed? Do we download the version number as specified in the file or just 'latest'?

We will always download the latest. If the `latest` version changes in between the version in the file being updated and running the Upgrade command, we will download and install the new latest. This has been brought up as something we might want to tweak, so that someone can download and install a specific version to avoid race conditions such as this, but we don't have that yet.

Reply:
> We download the package when running the Upgrade command.

But from the changes I see in the PR so far the 'download' code has been removed, not moved. Are you saying the actual call to 'download' is already in the Upgrade command code, and previously we were downloading twice (once for the check, once for the actual upgrade cmd)?

> This has been brought up as something we might want to tweak, so that someone can download and install a specific version to avoid race conditions such as this, but we don't have that yet.

..or at least show a warning/message that "the latest version is now n+1".